crypto/ec/ec_lib.c

```diff
@@ -1012,11 +1012,59 @@ int ec_group_simple_order_bits(const EC_GROUP *group)
     return BN_num_bits(group->order);
 }
 
+static int ec_field_inverse_mod_ord(const EC_GROUP *group, BIGNUM *r,
+                                    BIGNUM *x, BN_CTX *ctx)
+{
+    BIGNUM *e = NULL;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
+        goto err;
+
+    BN_CTX_start(ctx);
+    if ((e = BN_CTX_get(ctx)) == NULL)
+        goto err;
+
+    /* Check if optimized inverse is implemented */
+    if (group->mont_data != NULL) {
+        /*-
+         * We want inverse in constant time, therefore we utilize the fact
+         * order must be prime and use Fermats Little Theorem instead.
+         */
+        if (!BN_set_word(e, 2))
+            goto err;
+        if (!BN_sub(e, group->order, e))
+            goto err;
+        /*-
+         * Exponent e is public.
+         * No need for scatter-gather or BN_FLG_CONSTTIME.
+         */
+        if (!BN_mod_exp_mont(r, x, e, group->order, ctx, group->mont_data))
```

Comment:
To that end, e913d11f444e0b46ec1ebbf3340813693f4d869d was not really correct. When the exponent is public but only the base is secret, there's no need to use the slower exponent-protecting function. Although the special-case in `BN_mod_exp_mont` for `a` = 0 should probably be removed.

Reply:
> When the exponent is public but only the base is secret, there's no need to use the slower exponent-protecting function.

Good, so I'm no totally lost then :)

Reply:
> To that end, e913d11 was not really correct.

I hear you, but IMHO 8aed2a7548362e88e84a7feb795a3a97e8395008 is the better one to point at in this case. It's setting it on the exponent. Which is strange because, IIRC, both this FLT inversion and BN_mod_exp_mont_consttime are Intel contributions. I would expect them to know their own scatter-gather IPR :)